docs/sphinx/heps/hep2.rst

```diff
@@ -0,0 +1,310 @@
+HEP 2: The public API for |hpx| 2.0.0
+=====================================
+
+Motivation
+----------
+
+|hpx| has evolved to have a rather ad-hoc organization of functionality into
+namespaces, headers, and libraries. This HEP specifies the public API, and
+through which namespaces and headers the public API is accessible, for |hpx|
+2.0.0.
+
+Implementation
+--------------
+
+The public API is presented as
+
+- ``header name``
+  - ``hpx::x``
+  - ``hpx::y``
+
+where including ``header name`` gives you access to at least ``hpx::x`` and
+``hpx::y``. Functionality may be organized differently internally (both headers
+and namespaces).
+
+Functionality that corresponds to, or extends, standard library functionality,
+will be in the same namespace as in the standard library with ``std`` replaced
+by ``hpx``, and a header ``abc`` replaced by ``hpx/abc.hpp``. Functionality that
+is proposed for the standard library, and may thus be in ``std::experimental``
+and a header ``experimental/abc`` does not need to be in the
+``hpx::experimental`` namespace or a ``hpx/experimental`` header if the
+functionality is considered stable in |hpx|. Distributed extensions to local
+parallelism and concurrency will be in the ``hpx::distributed`` namespace and
+``hpx/distributed`` headers.
+
+Utilities
+.........
+
+- ``hpx/functional.hpp``
+  - ``hpx::function``
+  - ``hpx::unique_function``
+  - ``hpx::function_nonser``
+  - ``hpx::unique_function_nonser``
+  - ``hpx::bind``
+  - ``hpx::bind_front``
+  - ``hpx::bind_back``
+  - ``hpx::invoke``
+  - ``hpx::invoke_fused``
+  - ``hpx::result_of``
+  - more?
+
+- ``hpx/tuple.hpp``
+  - ``hpx::tuple``
+
+- ``hpx/any.hpp``
+  - ``hpx::any``
+
+- ``hpx/optional.hpp``
+  - ``hpx::optional``
+
+- ``hpx/format.hpp``
+  - ``hpx::format``?
+  - other utilities that go with ``format``
+
+- ``hpx/chrono.hpp``
+  - ``hpx::high_resolution_clock``?
+  - ``hpx::high_resolution_timer``?
+  - ``hpx::scoped_timer``?
+
+- ``hpx/exception.hpp``
+  - ``hpx::exception``
+  - ``HPX_THROW_EXCEPTION``
+
+- ``hpx/system_error.hpp``
+  - ``hpx::error_code``
+
+- ``hpx/assert.hpp``
+  - ``HPX_ASSERT``
+
+- ``hpx/serialization.hpp``
+  - what functionality actually needs to be public?
+  - Some classes have intrusive serialization in which case there's no need to
```

Comment:
"Some classes have intrusive serialization" Are you sure you don't mean "Some classes have implicit serialization..."

Reply:
No, intrusive, as in the serialization is implemented by member functions ("Some classes implement serialization intrusively" might be better, but I'm not planning on leaving that comment there). Using such a class gives you access to its serialization functionality whether you ask for it or not. Other classes have the serialization defined outside of the class (e.g. all the standard types) and including something which gives you the class does not necessarily give you the serialization functionality for it.